app_history.md

```diff
@@ -0,0 +1,423 @@
+
+- Provide a way for an application to reliably navigate through its own history stack.
+
+- Provide a reasonable layering onto and integration with the existing `window.history` API, in terms of spec primitives and ensuring non-terrible behavior when both are used.
+
+Non-goals:
+
+- Provide applications knowledge of cross-origin history entries or state.
+
+- Provide applications knowledge of other frames' entries or state.
+
+- Handle the case where the Android back button is being used as a "modal close signal"; instead, we believe that's best handled by [a separate API](./history_and_modals.md).
+
+- _TODO: something something allowing trapping users on the page; need to figure out which interceptions we allow._
+
+- Provide an elegant layering onto or integration with the existing `window.history` API. That API is quite problematic, and we can't be tied down by a need to make every operation in the new API isomorphic to one in the old API.
+
+_TODO: figure out which of [Chris's goals and non-goals](https://docs.google.com/document/d/1eL7tOQXVw7hyV7tudS7WukT9qIsmZm0o-qGxq0wP2Ko/edit#) we haven't incorporated, but should._
+
+## Proposal
+
+### The current entry, and single-page navigations
+
+The entry point for the app history API is `window.appHistory`. Let's start with `appHistory.currentEntry`, which is an instance of the new `AppHistoryEntry` class. This class has the following readonly properties:
+
+- `key`: a user-agent-generated UUID identifying this history entry. In the past, applications have used the URL as such a key, but the URL is not guaranteed to be unique.
+
+- `url`: the URL of this history entry (as a string).
+
+- `state`: returns the application-specific state stored in the history entry (or `null` if there is none).
+
+- `sameDocument`: a boolean indicating whether this entry is for the current document, or whether navigating to it will require a full navigation (either from the network, or from the browser's back/forward cache). Note: for `appHistory.currentEntry`, this will always be `true`.
+
+For single-page applications that want to update the current entry in the same manner as today's `history.replaceState()` APIs, we have the following:
+
+```js
+// Updates the URL shown in the address bar, as well as the url property.
+appHistory.updateCurrentEntry({ url });
+
+// Only updates the state property.
+appHistory.updateCurrentEntry({ state });
+
+// Update both at once.
+appHistory.updateCurrentEntry({ url, state });
+```
+
+_TODO: more realistic example, maybe something Redux-esque like `appHistory.updateCurrentEntry({ state: {...appHistory.currentEntry.state, newKey: newValue } })`._
+
+Similarly, to push a new entry in the same manner as today's `history.pushState()`, we have the following:
+
+```js
+// Pushes a new entry onto the app history list, copying the URL and state from the current one.
+await appHistory.pushNewEntry();
```

Comment:
I guess this also disposes all 'forward' history entries in the joint session history?